metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java

```diff
@@ -293,11 +300,49 @@ private void open() throws MetaException {
     boolean useFramedTransport = conf.getBoolVar(ConfVars.METASTORE_USE_THRIFT_FRAMED_TRANSPORT);
     int clientSocketTimeout = conf.getIntVar(ConfVars.METASTORE_CLIENT_SOCKET_TIMEOUT);
 
+    SocketFactory socketFactory = SocketFactory.getDefault();
+    Socket socket = null;
+    try {
+      socket = socketFactory.createSocket();
+      socket.setKeepAlive(true);
+
+      String hiveHost = conf.getVar(ConfVars.HIVE_SERVER2_THRIFT_BIND_HOST);
+      if (hiveHost.isEmpty()) {
+        UserGroupInformation loginUser = UserGroupInformation.getLoginUser();
+        if (loginUser != null && loginUser.hasKerberosCredentials()) {
+          hiveHost = SecurityUtil.getHostFromPrincipal(loginUser.getUserName());
+        }
+      }
+
+      if (!hiveHost.isEmpty()) {
+        InetAddress localAddr = NetUtils.getLocalInetAddress(hiveHost);
+        if (localAddr != null) {
+          socket.bind(new InetSocketAddress(localAddr, 0));
```

Comment:
I'm wondering if it would be useful to add an "else" here with a log message, because if the thrift host is configured or there's a krb principal, but there isn't a local address, then something is probably wrong in the config and it will be useful to point this out.

As you point out, it will use the default address, and will probably work, but the user will get confused if the address isn't what he expects.